docs/rfc/0003-collaborative-objects.md

```diff
@@ -0,0 +1,649 @@
+every remote we have and collect the change files for each object ID.
+
+At this point we have the hash linked graph of automerge changes, but we need
+to make sure that the merged document is authenticated and valid with respect
+to its schema. To do this we start at the roots of the hash graph and walk
+down the tree. As we encounter each change we check its signature, apply it
+and check that the new document does not violate the schema. If it does violate
+the schema we discard the change and all dependent changes. Finally, we have an
+authenticated document which respects a given schema.
+
+It is important to note that this merging is at this point not stored in the
+repository - it can be performed in memory and may be cached. The result is
+that the user sees a single merged view of the object based on the contents
+of the remotes they have replicated. That is, there is no additional
+merge-then-commit step.
+
+### Fetching Collaborative Objects
+
+Each time a repository creates a collaborative object tree it creates a ref
+pointing to that object at `refs/namespaces/<project>/rad/collaborative-objects/<typename>/<object ID>`, 
+where `object ID` is a unique identifier generated at creation time. This is
+the object ID in the `ID` field of the manifest.
+
+This allows us to fetch subsets of collaborative objects by specifying refspecs
+that match them. The downside is that we are adding a ref to the initial
+advertised refs, each of theses refs is around 250 bytes. If we consider a popular
+repository such as https://github.com/facebook/react/ you can see that they have
+nearly 10,000 issues (including both Open and Closed, which we must). That would 
+mean that the initial ref advertisment in any replication of this repository
+would be ~2.5Mb. If we could use v2 of the git pack protocol this problem would
+go away but alas, we must use v1. 
+
+To get around the problem we can add a parameter to the radicle URL which
+indicates either a single object ID or a type name which we wish to fetch
+from the remote, this would then be passed as a custom header to the transport
+and interpreted on the server. In this manner we can allow clients to choose
+when they want to replicate collaborative objects, which would allow staged
+fetches where we first fetch the repository and identities, and then fetch 
+collaborative objects.
+
+Therefore we add two parameters to a radicle URN:
+
+- `collab_object_id`: specifies that a server should only consider references to the given object ID
+- `collab_type_name`: specifies that a server should only consider references to the given type name
+
+These parameters are emitted by the radicle p2p transport as headers in a 
+similar fashion to the `n=` parameter for the nonce or the `ls=true`
+parameter for selecting the `upload-pack-ls` service. The Radicle git server
+can then use these parameters to filter the refs it operates on.
```

Comment:
It is potentially the right idea to add parameters to a _URL_, but the mechanics for the _URN_ are a bit different: essentially, every `URN/PeerId` pair is a git remote, and whenever you tell link to get a URN, you're creating a (persistent) remote. To restrict which branches are fetched, git's standard mechanism is a _fetch spec_ (the `git.<remote name>.fetch` key). This is something we can consider, but note that it has an effect on availability of the data (you either replicate those objects, or you don't).

Of interest is a way to "peek" at potential updates / contributions from peers which are _not_ in your tracking graph. This can be done semi-efficiently with v2 (which we'll hopefully be able to use soon) by just issuing a `ls-refs` with a `ref-prefix`. We need a place to store-and-forward this information, though.

Reply:
My intent with these parameters was to avoid advertising all the refs of the collaborative objects unless the peer wants them. As I understand protocol V1 the fetch spec approach does not avoid this right?

Reply:
We don't have to worry about v1 anymore -- I received a walkthrough from @Byron yesterday, and what I saw is sufficient to start ripping out libgit2 for network access NAO.

What we _should_ consider, however, is to make the refs paths [ref-prefix](https://git-scm.com/docs/protocol-v2#_ls_refs) friendly. Since we want `[remotes/*/]rad/{id,ids/*,signed_refs}` for replication (where we know the values of `remotes/*`), it might be worth considering to put the collaborative objects under their own namespace, eg. `cob/<type>/<uuid>`.

Reply:
Exciting! I'll rework this in that light then. Having a top level `cob/type/<uuid>` makes sense for prefixing purposes.